optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -258,6 +259,31 @@
             Function<A, GroupKeyA_> groupKeyAMapping, Function<A, GroupKeyB_> groupKeyBMapping,
             UniConstraintCollector<A, ResultContainer_, Result_> collector);
 
+    /**
+     * Combines the semantics of {@link #groupBy(Function, Function)} and {@link #groupBy(UniConstraintCollector)}.
+     * That is, the first and second facts in the tuple follow the {@link #groupBy(Function, Function)} semantics.
+     * The third fact is the result of applying the first {@link UniConstraintCollector#finisher()} on all the tuples
+     * of the original {@link UniConstraintStream} that belong to the group.
+     * The fourth fact is the result of applying the second {@link UniConstraintCollector#finisher()} on all the tuples
+     * of the original {@link UniConstraintStream} that belong to the group
+     * @param groupKeyAMapping never null, function to convert the original tuple into a first fact
+     * @param groupKeyBMapping never null, function to convert the original tuple into a second fact
+     * @param collectorC never null, the collector to perform the first grouping operation with
+     * @param collectorD never null, the collector to perform the first grouping operation with
+     * @param <GroupKeyA_> the type of the first fact in the destination {@link QuadConstraintStream}'s tuple
+     * @param <GroupKeyB_> the type of the second fact in the destination {@link QuadConstraintStream}'s tuple
+     * @param <ResultContainerC_> the mutable accumulation type (often hidden as an implementation detail)
+     * @param <ResultC_> the type of the third fact in the destination {@link QuadConstraintStream}'s tuple
+     * @param <ResultContainerD_> the mutable accumulation type (often hidden as an implementation detail)
+     * @param <ResultD_> the type of the fourth fact in the destination {@link QuadConstraintStream}'s tuple
+     * @return never null
+     */
+    <GroupKeyA_, GroupKeyB_, ResultContainerC_, ResultC_, ResultContainerD_, ResultD_>
```

Comment:
This is a good reason to think long and hard if we want to support PentaConstraintStreams.

QuadsStreams are still needed, definitely, despite this. It's worth it. Absolutely.

Reply:
My thinking on this:

- The number of use cases for `n`-streams will be decreasing as `n` is increasing. Therefore when (if) we do penta, hexa will bring even more diminishing results.
- This could all be solved with dynamic cardinality streams, that can handle any number of parameters.
- However, that would not be done with Java generics. In this kind of work, generics is the last thing that keeps me sane. If all I had throughout constraint streams were `Object`s, I would have gone mad.

All in all, if we have to add more streams, I'd much rather go for Penta than for Dynamic.

Reply:
One more thing to note - with dynamic cardinality streams not only would Lukas lose the type safety, but users as well.

Reply:
At some point, we will need to support dynamic cardinality streams. We can argue whether that's at Penta, Hexa, septi, octo, novem, ..., novemnonagint, centi or higher ;) As indeed users lose type safetly.
As for the design, I was thinking of `PolyConstraintStream extends AbstractConstraintStream<Object[]>`. Either way: out of scope for now.

Reply:
Just think about it for a second.
`NovemonagintConstraintStream<A, B, C, D, E, F, G, ... 99 in total>`

Reply:
@ge0ffrey The biggest pain seems to be the combinatorial explosion of grouping operations => too many possible groupKeys and collectors that bring you to/from stream of a different arity.